cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh

```diff
@@ -0,0 +1,188 @@
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__MEMORY_RESOURCE_ANY_RESOURCE2_H
+#define _CUDAX__MEMORY_RESOURCE_ANY_RESOURCE2_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// If the memory resource header was included without the experimental flag,
+// tell the user to define the experimental flag.
+#if defined(_CUDA_MEMORY_RESOURCE) && !defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+#  error "To use the experimental memory resource, define LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE"
+#endif
+
+// cuda::mr is unavable on MSVC 2017
+#if defined(_CCCL_COMPILER_MSVC_2017)
+#  error "The any_resource header is not supported on MSVC 2017"
+#endif
+
+#if !defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+#  define LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE
+#endif
+
+#include <cuda/__memory_resource/get_property.h>
+#include <cuda/__memory_resource/properties.h>
+#include <cuda/__memory_resource/resource.h>
+#include <cuda/std/__concepts/__concept_macros.h>
+#include <cuda/std/__utility/forward.h>
+#include <cuda/std/optional>
+
+#include <cuda/experimental/__utility/basic_any.cuh>
+
+namespace cuda::experimental::mr
+{
+using __property_value_or_void _CCCL_NODEBUG_ALIAS = //
+  _CUDA_VSTD::__type_try_catch< //
+    _CUDA_VSTD::__type_indirect_quote<__property_value_t>, // try to get the property value
+    _CUDA_VSTD::__type_always<void>>; // if it fails, return void
+
+template <class _Property>
+using __property_result_t _CCCL_NODEBUG_ALIAS = _CUDA_VSTD::__type_call1<__property_value_or_void, _Property>;
+
+template <class _Property>
+struct __with_property
+{
+  template <class _Ty>
+  _CUDAX_PUBLIC_API static auto __get_property(const _Ty& __obj, _Property __prop) noexcept //
+    -> __property_result_t<_Property>
+  {
+    static_assert(noexcept(get_property(__obj, __prop)));
+    if constexpr (!_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+    {
+      return get_property(__obj, __prop);
+    }
+  }
+
+  template <class...>
+  struct __iproperty : interface<__iproperty>
+  {
+    _CUDAX_API friend auto get_property(const __iproperty& __obj, _Property __prop) noexcept //
+      -> __property_result_t<_Property>
+    {
+      if constexpr (!_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+      {
+        return __cudax::virtcall<&__get_property<__iproperty>>(&__obj, __prop);
+      }
```

Comment:
Is that really necessary? It might actually give grief with compilers that mess up `if constexpr`
```suggestion
      return __cudax::virtcall<&__get_property<__iproperty>>(&__obj, __prop);
```

Reply:
which compilers mess up `if constexpr`? we really don't want to make an indirect function call if we know statically that it's a no-op.

Reply:
I believe NVCC before 12.0 will most likely complain about missing return statement because it has faulty if constexpr handling. You will need at least a `_CCCL_UNREACHABLE()` at the end

Reply:
but it isn't unreachable. when the `if constexpr` branch isn't taken, the function returns `void`, so falling off the end if the right thing to do.